gnomad_qc/v4/annotations/insilico_predictors.py

```diff
@@ -159,6 +159,114 @@ def import_spliceai_vcf(path: str) -> hl.Table:
+        :return: Hail Table with Pangolin scores.
+        """
+        ht = hl.import_vcf(
+            vcf_path,
+            min_partitions=1000,
+            force_bgz=True,
+            reference_genome="GRCh38",
+            skip_invalid_loci=True,
+        ).rows()
+        return ht
+
+    ht_g = import_pangolin_vcf(v4_genomes)
+    ht_e = import_pangolin_vcf(v4_exomes)
+
+    logger.info(
+        "Number of rows in genomes Pangolin HT: %s; "
+        "Number of rows in raw exomes Pangolin HT: %s. ",
+        ht_g.count(),
+        ht_e.count(),
+    )
+
+    ht = ht_g.union(ht_e)
+
+    logger.info("Exploding Pangolin scores...")
+    # `explode` will eliminate rows with empty array
+    # The VCF INFO of Pangolin is a string with the following format:
+    # gene1|pos_splice_gain:largest_increase|pos_splice_loss:largest_decrease|gene2...
+    # for example:
+    # Pangolin=ENSG00000121005.9|-86:0.25|38:-0.49|Warnings:||ENSG00000254238.1|-40:0.01|30:-0.17|Warnings:
+
+    ht = ht.annotate(pango=ht.info.Pangolin[0].split(delim="\|\|"))
+    ht = ht.explode(ht.pango)
+    logger.info("Number of rows in exploded Pangolin Hail Table: %s", ht.count())
+
+    logger.info("Annotating Pangolin scores...")
+    ht = ht.annotate(
+        pangolin=hl.struct(
+            delta_scores=(
+                hl.empty_array(hl.tfloat64).append(
+                    hl.float(ht.pango.split(delim=":|\\|")[2])
+                )
+            ).append(hl.float(ht.pango.split(delim=":|\\|")[4])),
+        )
+    )
+
+    # Using > instead of >= in case where delta score of splice gain and splice
+    # loss are the same but different sign, we will keep the positive score of
+    # splice gain
+    ht = ht.annotate(
+        largest_ds_gene=hl.if_else(
+            hl.abs(hl.min(ht.pangolin.delta_scores))
+            > hl.abs(hl.max(ht.pangolin.delta_scores)),
+            hl.min(ht.pangolin.delta_scores),
+            hl.max(ht.pangolin.delta_scores),
+        )
+    )
+    # TODO: tried hl.zip to make it one-step, but it's not working
+    ht = ht.select(ht.largest_ds_gene)
+    ht = ht.collect_by_key()
+    logger.info(
+        "Number of rows in Pangolin Hail Table after collect_by_key: %s", ht.count()
+    )
+    ht = ht.select(
+        pangolin=hl.struct(
+            largest_ds=hl.if_else(
+                hl.abs(hl.min(ht.values.largest_ds_gene))
+                > hl.abs(hl.max(ht.values.largest_ds_gene)),
+                hl.min(ht.values.largest_ds_gene),
+                hl.max(ht.values.largest_ds_gene),
+            )
+        )
+    )
+    logger.info(
+        "Number of variants indicating splice gain: %s;"
+        "Number of variants indicating splice loss: %s; "
+        "Number of variants with no splicing consequence: %s",
+        ht.filter(ht.pangolin.largest_ds > 0).count(),
+        ht.filter(ht.pangolin.largest_ds < 0).count(),
+        ht.filter(ht.pangolin.largest_ds == 0).count(),
```

Comment:
An single agg may be faster, I imaigne 3 filters + count are fairly slow?

Reply:
It's not very slow, but what agg do you mean? I found count_where can also only do one each time.